### PR TITLE
[Experimental] Preserve speculative decoding during prefill across DP ranks

### PR DESCRIPTION
**Pull request by Oasis-Git**
## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #35541405518](https://github.com/sgl-project/sglang/actions/runs/35541405518)
Latest PR Test (Extra): :x: [Run #35541405355](https://github.com/sgl-project/sglang/actions/runs/35541405355)
Latest PR Test (AMD ROCm 10): :x: [Run #35541405500](https://github.com/sgl-project/sglang/actions/runs/35541405500)

### CODE REVIEW

**Comment by Oasis-Git** [comment]
Base CI pass: https://github.com/sgl-project/sglang/actions/runs/35391417429

**Comment by Oasis-Git** [comment]
Extra CI Pass: https://github.com/sgl-project/sglang/actions/runs/35391417165

File: python/sglang/srt/managers/scheduler.py [hunk 1]
```diff
@@ -3720,6 +3720,7 @@ def get_next_batch_to_run(
             need_mlp_sync
             and not self.spec_algorithm.is_none()
             and not get_spec().speculative_skip_dp_mlp_sync
+            and not envs.SGLANG_ENABLE_DP_SPEC_PREFILL_COORDINATION.get()
```
Comment (ch-wan) [inline python/sglang/srt/managers/scheduler.py:3723]: **[bug]** `SGLANG_ENABLE_DP_SPEC_PREFILL_COORDINATION` skips the spec+DP mix guard for every algorithm. The per-phase token plan is only applied inside `EAGLEWorkerV2`, and that class rejects anything other than plain EAGLE. Multi-layer EAGLE (still `SpeculativeAlgorithm.EAGLE`, different worker), DFLASH, NGRAM, and frozen-KV MTP therefore schedule a heterogeneous step with no plan. Each rank then runs `spec_scale_global_num_tokens` with its own `num_tokens_per_req`, so a prefill rank's already-token counts and a decode rank's request counts disagree and the MLP collective sizes diverge. `--speculative-skip-dp-mlp-sync` is the existing form of this skip and is rejected unless the algorithm string is `EAGLE`; this env is not.

**Suggestion:** Skip the mix guard only when the live worker will actually install the plan, and reject the env at startup for every other algorithm the way `--speculative-skip-dp-mlp-sync` already does.

File: python/sglang/srt/speculative/eagle_worker_v2.py [hunk 2]
```diff
@@ -1308,6 +1311,16 @@ def __init__(
             get_spec().speculative_algorithm
         )
 
+        self.enable_dp_spec_prefill_coordination = (
+            envs.SGLANG_ENABLE_DP_SPEC_PREFILL_COORDINATION.get()
+        )
+        # Coordination for other speculative algorithms is not implemented yet.
+        if (
+            self.enable_dp_spec_prefill_coordination
+            and self.speculative_algorithm != SpeculativeAlgorithm.EAGLE
```
Comment (ch-wan) [inline python/sglang/srt/speculative/eagle_worker_v2.py:1320]: **[bug]** The guard compares identity with `SpeculativeAlgorithm.EAGLE`, so EAGLE3 crashes in `EAGLEWorkerV2.__init__` even though `create_worker` routes EAGLE3 through this same class and the same draft / verify / draft-extend widths. The registered DP-attention spec server is EAGLE3 (`--speculative-algorithm EAGLE3` in `test_eagle_dp_attention.py`). The phase plan does not depend on EAGLE versus EAGLE3; the comment that other algorithms are unimplemented does not apply here.

**Suggestion:** Accept EAGLE3 (`is_eagle3()`, or `is_eagle()` minus frozen-KV MTP, which has its own worker). Do not treat EAGLE3 as an unimplemented algorithm.

File: python/sglang/srt/speculative/dp_spec_prefill_coordination.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        tokens, logprobs = self.phase_counts(phase)
+        # EP-only batches retain only their local counts.
+        if len(batch.global_num_tokens) == 1:
+            tokens, logprobs = [tokens[rank]], [logprobs[rank]]
+        elif len(batch.global_num_tokens) != len(tokens):
+            raise ValueError("Unexpected DP synchronization group width")
+        batch.global_num_tokens = tokens
+        batch.global_num_tokens_for_logprob = logprobs
+        batch.dp_spec_prefill_coordination_applied = True
+        batch.is_extend_in_batch = True
```
Comment (ch-wan) [inline python/sglang/srt/speculative/dp_spec_prefill_coordination.py:54]: **[bug]** `apply` forces `is_extend_in_batch = True` on draft, target, and draft-extend. `prepare_mlp_sync_batch` uses that flag, together with MAX_LEN, to rewrite decode / target-verify / draft-extend / idle forwards into extend rows (`forward_batch_info.py` around the `is_extend_in_batch and dp_padding_mode.is_max_len()` branch). Draft-phase counts are 0 on every prefill rank, so `min(global_num_tokens) == 0`. Hybrid models set `dp.max_len_with_idle`, which selects MAX_LEN; the decode rank's spec info is a draft input, so the hybrid non-draft branch is skipped and the non-empty rank hits `assert seq_lens.shape[0] == 0`. pplx, and DeepEP-v2 with `SGLANG_DEEPEP_V2_FORCE_MAX_LEN`, always pick MAX_LEN and instead take the 1-token extend rewrite, so verify no longer has `draft_token_num` rows. SUM_LEN (non-hybrid, a2a none) does not take this branch; the token scaling itself matches `spec_scale_global_num_tokens` there.

**Suggestion:** Do not reuse `is_extend_in_batch` as the SUM_LEN switch. Keep decode/verify/draft forwards out of the fabricated-extend path, and select SUM_LEN explicitly for the coordinated step.

File: python/sglang/srt/speculative/eagle_worker_v2.py [hunk 4]
```diff
@@ -1557,6 +1547,129 @@ def forward_batch_generation(
...
+        with (
+            self.draft_worker.draft_tp_context(
+                self.draft_worker.draft_runner.tp_group,
+                owns_attention=self.draft_worker.draft_owns_attention,
+            ),
+            speculative_moe_backend_context(),
+            speculative_moe_a2a_backend_context(),
+            spec_stage_span("draft"),
+        ):
+            verify_input = self.draft_worker.draft(draft_batch)
```
Comment (ch-wan) [inline python/sglang/srt/speculative/eagle_worker_v2.py:1643]: **[bug]** The pure-decode prologue is not used on a heterogeneous step. `speculative_num_steps == 0` normally builds a 1-node verify input and can skip draft-extend under `SGLANG_SPEC_SKIP_ZERO_STEP_DRAFT_EXTEND` (adaptive spec reaches 0 steps). This path always calls `draft()`. On a non-idle rank `draft_forward`'s loop is `range(0)` and then indexes `token_list[0]`; the prefill rank's idle draft returns `None`s and builds an idle verify input, so the ranks do not even fail together. The same call also drops `pp_proxy_tensors` (the prefill branch forwards them) and never runs the PP tail-draft that the pure-decode branch runs after verify.

**Suggestion:** Share the decode prologue: `activate_step_by_batch`, the steps==0 trivial verify and draft-extend skip, and pass `pp_proxy_tensors` into `verify`. Run the PP tail-draft for non-prefill ranks when `pp_size > 1`.

File: python/sglang/srt/speculative/dp_spec_prefill_coordination.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
+"""Plan token counts for prefill and speculative decoding across DP ranks."""
+
+from dataclasses import dataclass
+from typing import List
+
+import torch
+
+
+@dataclass(frozen=True)
```
Comment (ch-wan) [inline python/sglang/srt/speculative/dp_spec_prefill_coordination.py:9]: **[suggestion]** `DPSpecPrefillCoordinationPlan` is a new `@dataclass`. New data containers in this repo are `msgspec.Struct` (frozen, kw-only as needed), not dataclasses.

**Suggestion:** Replace the dataclass with a frozen `msgspec.Struct` and keep `phase_counts` / `apply` as methods on it.

File: test/registered/spec/eagle/test_eagle_dp_attention.py [hunk 6]
```diff
@@ -25,7 +28,7 @@
 register_amd_ci(est_time=200, suite="stage-c-test-4-gpu-amd")
 
 
-class TestEAGLE3EngineDPAttention(CustomTestCase):
+class TestEAGLE3EngineDPAttention(DPSpecPrefillCoordinationKit, CustomTestCase):
```
Comment (ch-wan) [inline test/registered/spec/eagle/test_eagle_dp_attention.py:31]: **[suggestion]** `DPSpecPrefillCoordinationKit` is mixed into the EAGLE3 DP-attention server, but `setUpClass` never sets `SGLANG_ENABLE_DP_SPEC_PREFILL_COORDINATION`. With the flag off the scheduler still serializes prefill and decode, so the overlap, idle-transition, and abort cases pass on today's behavior. Setting the flag on this server would instead crash worker init (the EAGLE3 guard). The sleeps also do not force a heterogeneous step, and the CPU worker test mocks `draft` / `verify`, so it cannot catch the MAX_LEN rewrite or the steps==0 crash.

**Suggestion:** Launch a plain-EAGLE (or EAGLE3, once the worker accepts it) server with the env on, and add a case that fails if ranks disagree on per-phase `global_num_tokens` or if steps==0 still enters `draft()`.

File: test/registered/unit/spec/test_dp_spec_prefill_coordination.py [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+        global_num_tokens=list(counts),
+        global_num_tokens_for_logprob=list(logprobs),
+        dp_spec_prefill_coordination_applied=False,
+        is_extend_in_batch=False,
+        can_run_decode_cuda_graph=True,
+        can_run_dp_prefill_cuda_graph=True,
+    )
+
+
+class TestDPSpecPrefillCoordinationPlan(unittest.TestCase):
```
Comment (ch-wan) [inline test/registered/unit/spec/test_dp_spec_prefill_coordination.py:44]: **[nit]** Both new CPU test classes subclass `unittest.TestCase`. Registered tests subclass `CustomTestCase`.

**Suggestion:** Inherit `CustomTestCase`.

File: python/sglang/srt/speculative/eagle_worker_v2.py (hunk 2 above)
Comment (Oasis-Git) [reply python/sglang/srt/speculative/eagle_worker_v2.py]: solved

File: test/registered/unit/spec/test_dp_spec_prefill_coordination.py (hunk 7 above)
Comment (Oasis-Git) [reply test/registered/unit/spec/test_dp_spec_prefill_coordination.py]: solved

File: python/sglang/srt/speculative/dp_spec_prefill_coordination.py (hunk 5 above)
Comment (Oasis-Git) [reply python/sglang/srt/speculative/dp_spec_prefill_coordination.py]: solved

File: python/sglang/srt/managers/scheduler.py (hunk 1 above)
Comment (Oasis-Git) [reply python/sglang/srt/managers/scheduler.py:3723]: solved

File: python/sglang/srt/speculative/dp_spec_prefill_coordination.py (hunk 3 above)
Comment (Oasis-Git) [reply python/sglang/srt/speculative/dp_spec_prefill_coordination.py:51]: solved

**Comment by Oasis-Git** [comment]
/rerun-test test/registered/spec/eagle/test_dp_spec_prefill_coordination.py test/registered/unit/spec/test_dp_spec_prefill_coordination.py test/registered/unit/model_executor/test_mlp_sync_pad_unpad.py

File: test/registered/spec/eagle/test_eagle_dp_attention.py (hunk 6 above)
Comment (Oasis-Git) [reply test/registered/spec/eagle/test_eagle_dp_attention.py]: solved

**Review by ch-wan — COMMENTED** [review]
**Reviewed by agent.**

This commit closes the startup algorithm gate, the EAGLE3 reject, the MAX_LEN extend rewrite, the dataclass, and the test-base gaps. With the flag left off, scheduling and the forward path still match the previous behavior.

The heterogeneous path still always calls `draft()`, so `speculative_num_steps == 0` misses the trivial-verify prologue and `verify()` still drops `pp_proxy_tensors`. That existing thread is unchanged.

## Issue counts by severity

- bugs: 1
- suggestions: 0
- nits: 0

File: python/sglang/srt/speculative/eagle_worker_v2.py (hunk 4 above)
Comment (ch-wan) [reply python/sglang/srt/speculative/eagle_worker_v2.py:1647]: Still open on faa4f781. This commit does not consult speculative_num_steps before this draft() call, and verify() just below still omits pp_proxy_tensors. The steps==0 prologue remains only on the pure-decode arm of forward_batch_generation.
